### ISSUE TITLE
Empty (begin) evaluates to zero

### ISSUE THREAD

**Issue by lassik** [shirok/Gauche#514]
```
gosh> (eqv? (begin) 0)
#t
```

R7RS doesn't say anything clear about an empty begin but `0` is probably an unexpected return value. Some implementations return `#<undef>`; does Gauche have a void value? Since Gauche has real multiple-values, one option is that empty begin is equivalent to `(values)`.

**Comment by shirok** [comment]
(begin) compiles to "do nothing"---literally, do nothing, even not setting a particular value.  If you look at the result, you wee whatever remained in the VM regsiter.  In the above example, it just happens that VM register holds 0.  (When a procedure is called the accumulator of VM holds the number of arguments, so I guess that's what you're seeing here.)

If other value happens to be in the VM register you'll see that instead:

```
gosh> (begin 'a (begin))
a
```

Since `(begin)` is undefined, and loading `#<undef>` or something takes extra instructions, I'm reluctant to change the current behavior.  I can document this, though.

**Comment by lassik** [comment]
Wow, interesting! I had no idea to consider that perspective. Gauche is an impressively fast interpreter.

Documenting it sounds good. Is there an easy way for the compiler to check whether the return value of `begin` has a receiver? If there is, it could return a consistent value to tonly in that case; otherwise compile to no instructions.

**Comment by pclouds** [comment]
> Documenting it sounds good. Is there an easy way for the compiler to check whether the return value of begin has a receiver? If there is, it could return a consistent value only in that case; otherwise compile to no instructions.

I worked on a JIT compiler before that did something like this. It basically scanned ahead (until it hit a branch instruction or one that overwrites the same register) to see if the value is used. If not, don't even bother generate code to calculate the return value.

Your suggestion goes the opposite way of course. However if it turns out such "scan ahead and eliminate" strategy may result significant execution speedup for other cases and not cause compiler slowdown, then perhaps it's a good idea to do it.

No idea how complicated that would be though. Gauche compiler is still a mystery to me.

**Comment by shirok** [comment]
Actually, it's beacause we don't want to spend much time in compiling, which is done on-the-fly.  A straightforward way is to generate loading `#<undef>` by `(begin)`, then expect it to be removed in dead-code elimination pass.

**Comment by pclouds** [comment]
> A straightforward way is to generate loading #\<undef> by (begin), then expect it to be removed in dead-code elimination pass.

Does not seem to work. The change (I assume) is like this

```patch
diff --git a/src/compile-1.scm b/src/compile-1.scm
index ce9156b0b..e3f3c9a5d 100644
--- a/src/compile-1.scm
+++ b/src/compile-1.scm
@@ -1501,7 +1501,7 @@
 ;; Begin .....................................................
 
 (define-pass1-syntax (begin form cenv) :null
-  ($seq (imap (cut pass1 <> cenv) (cdr form))))
+  ($seq (imap (cut pass1 <> cenv) (cons '(undefined) (cdr form)))))
 
 ;; Lazy & Delay ..............................................
 
```

But after pass 5 it's still not eliminated

```
gosh$ ((with-module gauche.internal compile-p5) '(begin #t))
=== main_code (name=%toplevel, cc=0x7f8814d41a80, codevec=0x7f8814b0c9a0, size=3, const=0 stack=0):
signatureInfo: #f
     0 CONSTU                   ; (undefined)
     1 CONST-RET #t
```

This works, unsurprisingly since it singles out this case, but there's more computation for every `begin` form we have to handle

```patch
diff --git a/src/compile-1.scm b/src/compile-1.scm
index ce9156b0b..8c5d3f70f 100644
--- a/src/compile-1.scm
+++ b/src/compile-1.scm
@@ -1501,7 +1501,9 @@
 ;; Begin .....................................................
 
 (define-pass1-syntax (begin form cenv) :null
-  ($seq (imap (cut pass1 <> cenv) (cdr form))))
+  ($seq (imap (cut pass1 <> cenv) (if (null? (cdr form))
+                                    '((undefined))
+                                    (cdr form)))))
 
 ;; Lazy & Delay ..............................................
 
```

But maybe this just points out that we can improve the dead-code elimination optimization a bit (even if the new change is costly, we could turn it on in precompiler only; which still benefits us)

**Comment by shirok** [comment]
Yeah what I meant by expecting to be removed by DCE is that "when we have a good DCE pass".   We don't, now.

Conditionally generating `(undefined)` is a partial solution, since empty begin can occur in the middle of sequence (e.g. `(lambda () (begin) expr ...)`.   Especially as a result of macro expansion.

Unless this really becomes an issue, I'd rather put it off to DCE implementation.

**Comment by shirok** [comment]
Correction: `(undefined)` in the middle of sequence is eliminated during pass5.  So your latter suggestion may be effective.

**Comment by shirok** [comment]
Aargh, it doesn't work, for there's code that expects `(begin)` becomes nothing---e.g. after macro expansion it produces like `(begin expr (begin))`.  Such code is not RnRS conformant but the change can break existing code.

**Comment by pclouds** [comment]
Aha. Good reason to close this issue then?

**Comment by shirok** [comment]
Technically the issue is that we want to place #<undef> if and only if an entire expression becomes empty, and that's not addressed yet.  Let's keep this open and wait for some brilliant idea hits us.

**Comment by pclouds** [comment]
Off topic, I suppose you did consider at some point to make the precompiler generate C code instead of VM instructions and decided not to do it. After all you experimented with JIT and both approaches are not that different, at least they both avoid the interpreter loop. So what ware the reasons for not doing it?

In the above code, I suppose if we generate C code for updating VAL0 with #\<undefined> and #t, any mature C compiler should be able to eliminate the undefined assignment for us...

**Comment by shirok** [comment]
Actually I'm doing it.  It's not committed to the master branch.  The idea is that I have an alternative pass5 which first construct basic blocks for abstract register machine, then I can generate for various target (C and JIT is the primary target, but wasm would also be interesting.)

The last try of JIT was to substitute each VM instruction for a sequence of assembly.  It is still bound to Gauche VM architecture.  But we don't need to, and converting each basic block to assembly would have much less overhead, I expect.

**Comment by pclouds** [comment]
Slightly off topic again, but 

```
;;===============================================================
;; Pass 3.  Optimization stage 2
;;

;; Closure optimization can introduce superfluous $LET, which can
;; be optimized further.  (In fact, pass2 and pass3 can be repeated
;; until no further optimization can be possible.  However, compilation
;; speed is also important for Gauche, so we just run this pass once.)
```

Speed is not important to the precompiler however. Which means we could repeat passes 2 and 3 at least a couple more times in precompile mode (_if_ experiments show that it produces better code)?

**Comment by shirok** [comment]
Right, and that's the plan.  I think I've written a rough dead code elimination pass before (not the clever one, but the pass 5 peephole optimization leaves some obvious unused code and removing them make code size smaller), but now I don't remember where it'd gone... Anyway, precomp can do a lot more aggressive optimizations, for sure.

**Comment by pclouds** [comment]
> Which means we could repeat passes 2 and 3 at least a couple more times in precompile mode (if experiments show that it produces better code)?

Bad news. I used compiler.scm (from the benchmark suite in larceny) as compile input, ran it through `compile-p5` with passes 2/3 repeated twice. Identical results.

Just in case I made a mistake, the new `pass2-4` is

```scheme
(define-inline (pass2-4 iform module)
  (pass4 (pass3 (pass2 (pass3 (pass2 iform) #f)) #f) module))
```

and test code

```scheme
(use file.util)
((with-module gauche.internal compile-p5) (file->sexp-list "/tmp/compiler.scm"))
```

**Comment by pclouds** [comment]
Regarding native code compiler (through C), this link looks interesting because emacs is lispy and has a stack-based VM. It looks like they take stack-based byte code (Gauche's pass5), optimize a lot more then feed it to gcc.

They may have an interesting point of skipping C as an intermediate representation because a C compiler does not know many things about Lisp...

[Bringing GNU Emacs to native code](https://news.ycombinator.com/item?id=23066971)

**Comment by shirok** [comment]
Regarding skipping C---you're right, and that's why I'm looking at the JIT as the main goal.  However, the Gauche compiler and runtime themselves must be bootstrapped on a machine that doesn't have Gauche initially.  And translating to C is probably the easiest way.

Here, "translate to C" doesn't mean Scheme procedures map to C procedures.   It'll be more like a bunch of small code fragments called on a trampoline (If C code calls back to Scheme, we have to split the C code to two functions to use continuation-passing style).  I don't expect much speed boost except a bit from elimination of micro VM stack operations.

**Comment by pclouds** [comment]
> Actually I'm doing it. It's not committed to the master branch. The idea is that I have an alternative pass5 which first construct basic blocks for abstract register machine, then I can generate for various target (C and JIT is the primary target, **but wasm would also be interesting**.)

Not exactly related to the compiler. But I was curious and just trying out. It looks like it's not that hard to build a very minmal gauche with emscripten. There's a problem in bdwgc (so not really practical yet). But if I disable it, then at least I can evaluate some expressions from the browser :)
